### PR TITLE
add arg `padding_free` to DataCollatorForCompletionOnlyLM

### PR DESCRIPTION
**Pull request by RhuiDih**
add argument `DataCollatorForCompletionOnlyLM(padding_free=True)` to utilize https://github.com/huggingface/transformers/pull/31629

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @kashif this is linked to https://github.com/huggingface/transformers/pull/31629 which will be in the next transformers release!

**Comment by kashif** [comment]
thanks @ArthurZucker

**Comment by pcuenca** [comment]
Gentle ping and info, this post covers the support in transformers: https://huggingface.co/blog/packing-with-FA2

**Review by kashif — APPROVED** [review]

File: tests/test_data_collator_completion_only.py [hunk 1]
```diff
@@ -104,3 +104,37 @@ def test_data_collator_handling_of_long_sequences(self):
...
+        if tokenizer.pad_token_id is None:
+            tokenizer.pad_token = tokenizer.eos_token
+            tokenizer.pad_token_id = tokenizer.eos_token_id
+            
+        inst1 = "### System: You are a helpful assistant.\n\n### User: How much is 2+2?\n\n### Assistant: 2+2 equals 4"
+        inst2 = "### System: You are a honest and helpful assistant.\n\n### User: What is the answer of 22x22?\n\n### Assistant: 22x22 equals 484"
+
+        response_template = "\n### Assistant:"
+        collator = DataCollatorForCompletionOnlyLM(response_template, tokenizer=tokenizer)
+        collator_paddingfree = DataCollatorForCompletionOnlyLM(response_template, tokenizer=tokenizer, padding_free=True)
```
Comment (kashif) [inline tests/test_data_collator_completion_only.py:119]:
```suggestion
        collator_paddingfree = DataCollatorForCompletionOnlyLM(
            response_template, tokenizer=tokenizer, padding_free=True
        )
```

File: tests/test_data_collator_completion_only.py [hunk 2]
```diff
@@ -104,3 +104,37 @@ def test_data_collator_handling_of_long_sequences(self):
...
+
+        tokenized_instruction = [tokenizer(x, add_special_tokens=False) for x in [inst1, inst2]]
+        batch = collator(tokenized_instruction)
+        batch_paddingfree = collator_paddingfree(tokenized_instruction)
+
+        assert not "attention_mask" in batch_paddingfree
```
Comment (kashif) [inline tests/test_data_collator_completion_only.py:125]:
```suggestion
        assert "attention_mask" not in batch_paddingfree
```

File: tests/test_data_collator_completion_only.py [hunk 3]
```diff
@@ -104,3 +104,37 @@ def test_data_collator_handling_of_long_sequences(self):
...
+        assert "input_ids" in batch_paddingfree and "labels" in batch_paddingfree and "position_ids" in batch_paddingfree
```
Comment (kashif) [inline tests/test_data_collator_completion_only.py:126]:
```suggestion
        assert (
            "input_ids" in batch_paddingfree and "labels" in batch_paddingfree and "position_ids" in batch_paddingfree
        )
```

File: tests/test_data_collator_completion_only.py [hunk 4]
```diff
@@ -104,3 +104,37 @@ def test_data_collator_handling_of_long_sequences(self):
...
+        assert batch_paddingfree["input_ids"].size() == batch_paddingfree["labels"].size() == batch_paddingfree["position_ids"].size()
```
Comment (kashif) [inline tests/test_data_collator_completion_only.py:127]:
```suggestion
        assert (
            batch_paddingfree["input_ids"].size()
            == batch_paddingfree["labels"].size()
            == batch_paddingfree["position_ids"].size()
        )
```

File: trl/trainer/utils.py [hunk 5]
```diff
@@ -211,6 +213,14 @@ def torch_call(self, examples: List[Union[List[int], Any, Dict[str, Any]]]) -> D
                 if len(response_token_ids_idxs) < len(human_token_ids_idxs):
                     batch["labels"][i, human_token_ids_idxs[-1] :] = self.ignore_index
 
+        if self.padding_free:
+            # remove padding, `attention_mask` and add `position_ids`
+            attn_mask = batch.pop("attention_mask")
+            batch["input_ids"] = batch["input_ids"][attn_mask.bool()].unsqueeze(0)
+            batch["position_ids"] = attn_mask.cumsum(1)[attn_mask.bool()].unsqueeze(0) - 1
+            batch["labels"] = batch["labels"][attn_mask.bool()].unsqueeze(0) 
+            batch["labels"][batch["position_ids"]==0] = self.ignore_index
```
Comment (kashif) [inline trl/trainer/utils.py:222]:
```suggestion
            batch["labels"] = batch["labels"][attn_mask.bool()].unsqueeze(0)
            batch["labels"][batch["position_ids"] == 0] = self.ignore_index
```

File: tests/test_data_collator_completion_only.py [hunk 6]
```diff
@@ -104,3 +104,45 @@ def test_data_collator_handling_of_long_sequences(self):
         encoded_instance = self.collator.torch_call([self.tokenized_instruction])
         result = torch.all(encoded_instance["labels"] == -100)
         assert result, "Not all values in the tensor are -100."
+
+    def test_padding_free(self):
+        tokenizer = AutoTokenizer.from_pretrained("trl-internal-testing/dummy-GPT2-correct-vocab")
+        if tokenizer.pad_token_id is None:
+            tokenizer.pad_token = tokenizer.eos_token
+            tokenizer.pad_token_id = tokenizer.eos_token_id
+
```
Comment (kashif) [inline tests/test_data_collator_completion_only.py:113]:
```suggestion

```

File: tests/test_data_collator_completion_only.py [hunk 7]
```diff
@@ -104,3 +104,44 @@ def test_data_collator_handling_of_long_sequences(self):
...
+        collator = DataCollatorForCompletionOnlyLM(response_template, tokenizer=tokenizer)
+        collator_paddingfree = DataCollatorForCompletionOnlyLM(
+            response_template, tokenizer=tokenizer, padding_free=True
+        )
+
+        tokenized_instruction = [tokenizer(x, add_special_tokens=False) for x in [inst1, inst2]]
+        batch = collator(tokenized_instruction)
+        batch_paddingfree = collator_paddingfree(tokenized_instruction)
+
+        assert "attention_mask" not in batch_paddingfree
```
Comment (lewtun) [inline tests/test_data_collator_completion_only.py:126]: Can you please replace these assert statements with assert methods from `unittest`, e.g. `self.assertTrue()` for this case

**Review by lewtun — COMMENTED** [review]
Thanks for adding this neat extension to the collator! It LGTM barring the unit tests using the assert methods from `unittest`

File: tests/test_data_collator_completion_only.py (hunk 7 above)
Comment (kashif) [inline tests/test_data_collator_completion_only.py:126]:
```suggestion
        self.assertNotIn("attention_mask", batch_paddingfree)
```

Comment (RhuiDih) [reply tests/test_data_collator_completion_only.py]: @kashif @lewtun thanks for the suggestion, just pushed the update

File: tests/test_data_collator_completion_only.py [hunk 8]
```diff
@@ -104,3 +104,47 @@ def test_data_collator_handling_of_long_sequences(self):
...
+        self.assertIn("labels", batch_paddingfree)
+        self.assertIn("position_ids", batch_paddingfree)
+        self.assertEqual(
+            batch_paddingfree["input_ids"].size(),
+            batch_paddingfree["labels"].size()
+        )
+        self.assertEqual(
+            batch_paddingfree["labels"].size(),
+            batch_paddingfree["position_ids"].size()
+        )
```
Comment (pcuenca) [inline tests/test_data_collator_completion_only.py:137]:
```suggestion
        self.assertEqual(batch_paddingfree["input_ids"].size(), batch_paddingfree["labels"].size())
        self.assertEqual(batch_paddingfree["labels"].size(), batch_paddingfree["position_ids"].size())
```

Comment (pcuenca) [reply tests/test_data_collator_completion_only.py]: Make formatter happy

**Pull request merged**